doc/designs/quic-design/connection-id-cache.md

```diff
@@ -0,0 +1,251 @@
+short form packets include the connection ID but no length byte.  Random
+connection IDs of this length will be allocated.  Note that no additional
+information will be contained in the connection ID.
+
+There will be one global set of local connection IDs, `QUIC_ROUTE_TABLE`,
+which is shared by all connections over all SSL_CTX objects.  This is
+used to dispatch incoming datagrams to their correct destination and
+will be implemented as a dictionary.
+
+### Notes
+
+* For MVP, it would be sufficient to only use a zero length connection ID.
+* For MVP, a connection ID to QUIC_CONNECTION mapping need not be implemented.
+* Post MVP, funnelling all received packets through a single socket is
+  likely to be bottleneck.
+  An alternative would be receiving from <host address, source port> pairs.
+* For MVP, the local connection ID cache need only have one element.
+  I.e. there is no requirement to implement any form of lookup.
+
+Routes
+------
+
+A pair of connection IDs identify a route between the two ends of the
+communication.  This contains the connection IDs of both ends of the
+connection and the common sequence number.  We need to be able to:
+
+* select a connection ID by sequence number and retire that and all older IDs
+  and
+* select a connection ID by sequence number and drop that and all older IDs.
+
+It is likely that operations on local and remote connection IDs can be
+subsumed by the route functionality.
+
+API
+---
+
+QUIC connection IDs are defined in #18949 but some extra functions
+are available:
+
+```c
+/* QUIC connection ID representation. */
+#define QUIC_MAX_CONN_ID_LEN   20
+
+typedef struct quic_conn_id_st {
+    unsigned char id_len;
+    unsigned char id[QUIC_MAX_CONN_ID_LEN];
+} QUIC_CONN_ID;
+
+static ossl_unused ossl_inline int ossl_quic_conn_id_eq(const QUIC_CONN_ID *a,
+                                                        const QUIC_CONN_ID *b);
+
+/* New functions */
+int ossl_quic_conn_id_set(QUIC_CONN_ID *cid, unsigned char *id,
+                          unsigned int id_len);
+
+int ossl_quic_conn_id_generate(QUIC_CONN_ID *cid);
+```
+
+### Remote Connection ID APIs
+
+```c
+typedef struct quic_remote_conn_id_cache_st QUIC_REMOTE_CONN_ID_CACHE;
+
+/*
+ * Allocate and free a remote connection ID cache
+ */
+QUIC_REMOTE_CONN_ID_CACHE *ossl_quic_remote_conn_id_cache_new_conn(
+        unsigned int id_limit   /* [active_connection_id_limit] */
+    );
+void ossl_quic_remote_conn_id_cache_free(QUIC_REMOTE_CONN_ID_CACHE *cache);
+
+/*
+ * Add a remote connection ID to the cache
+ */
+int ossl_quic_cache_add_remote_conn_id(QUIC_REMOTE_CONN_ID_CACHE *cache,
+                                       const QUIC_CONNECTION *conn,
+                                       const unsigned char *conn_id,
+                                       size_t conn_id_len,
+                                       uint64_t seq_no);
+
+/*
+ * Query a remote ID for a connection.
```

Comment:
Not sure I understand what this query is doing? Can you elaborate on what this means?

Reply:
It gets a remote connection ID associated with a connection, so something can be sent.  Multiple routes are allowed, so there are potentially multiple IDs available.

However, host/port pairs also need to be connected to the ID somehow (so the packet can be sent correctly).

Reply:
Clarified